File: 2024/d6.py
from pprint import pprint

def get_next_index(cursor, row, col):
    match cursor:
        case '^':
            return (row-1, col)
        case '>':
            return (row, col+1)
        case 'v':
            return (row+1, col)
        case '<':
            return (row, col-1)

data = []
with(open('./data/day6.txt') as file):
    row = 0
    col = -1
    line_ctr = 0
    while (line:= file.readline()):
        line = line.strip()
        data.append(list(line))
        if (idx:=line.find('^'))!= -1: row, col = line_ctr, idx
        line_ctr += 1
start_row = row
start_col = col

direction = 0
directions = ['^', '>', 'v', '<']
while(True):
    cursor = data[row][col]
    if not isinstance(cursor, str):
        cursor = cursor[-1]
    (next_row, next_col) = get_next_index(cursor, row, col)

    if not (0 <= next_row < len(data) and 0 <= next_col < len(data[next_row])): break
    next_tile = data[next_row][next_col]
    if not isinstance(next_tile, str):
        next_tile = next_tile[-1]
    match next_tile:
        case '.':
            if isinstance(data[row][col], list):
                data[row][col].append(cursor)
            else:
                data[row][col] = cursor
            data[next_row][next_col] = cursor
            row = next_row
            col = next_col
        case '#':
            direction = (direction + 1) % 4
            if isinstance(data[row][col], list):
                data[row][col].append(directions[direction])
            else:
                data[row][col] = [directions[direction]]
        case _:
            if isinstance(data[next_row][next_col], list):
                data[next_row][next_col].append(cursor)
            else:
                data[next_row][next_col] = [next_tile, cursor]
            row = next_row
            col = next_col

row = start_row
col = start_col
obstruction_ctr = 0
direction = 0
cursor = '^'
while(True):
    print(row, col)
    (next_row, next_col) = get_next_index(cursor, row, col)
    if not (0 <= next_row < len(data) and 0 <= next_col < len(data[next_row])): break
    next_tile = data[next_row][next_col]
    match next_tile:
        case '#':
            print(next_tile)
            direction = (direction + 1) % 4
            cursor = directions[direction]
        case _:
            new_cursor = directions[(direction + 1) % 4]
            if new_cursor in data[row][col] or :
                obstruction_ctr += 1
            row = next_row
            col = next_col


print(obstruction_ctr)

File: 2024/d6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    FILE *filePtr = fopen("./data/day6.txt", "r");

    // Get file size 
    fseek(filePtr, 0, SEEK_END); 
    int size = ftell(filePtr); 
    fseek(filePtr, 0, SEEK_SET);

    // Read data
    char *rawData = (char*) calloc (size, sizeof(char));	
    fread(rawData, sizeof(char), size, filePtr);
    
    char *data[131];
    char *line;
    line = strtok(rawData,"\n");
    int ctr = 0;
    while(line != NULL)
    {
        data[ctr] = line;
        line = strtok(NULL,"\n");
        ctr++;
    }

    int rowIndex = -1;
    int colIndex = -1;
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < strlen(data[i]); j++) {
            if (data[i][j] != '.' && data[i][j] != '#') {
                rowIndex = i;
                colIndex = j;
                break;
            }
        }
        if (colIndex != -1) {
            break;
        }
    }
    int startingRow = rowIndex;
    int startingCol = colIndex;

    // while (1) {
    //     int nextRowIndex = rowIndex;
    //     int nextColIndex = colIndex;
    //     switch (data[rowIndex][colIndex]) {
    //     case '^':
    //         nextRowIndex--;
    //         break;
    //     case '>':
    //         nextColIndex++;
    //         break;
    //     case 'v':
    //         nextRowIndex++;
    //         break;
    //     case '<':
    //         nextColIndex--;
    //         break;
    //     default:
    //         break;
    //     }

    //     if ((0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) {
    //         nextTile = data[nextRowIndex][nextColIndex];
    //     }
    //     switch (nextTile) {
    //     case '.':
    //         cursor = data[rowIndex][colIndex];
    //         distinctTiles++;
    //         data[rowIndex][colIndex] = 'X';
    //         if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
    //         data[nextRowIndex][nextColIndex] = cursor;
    //         rowIndex = nextRowIndex;
    //         colIndex = nextColIndex;
    //         break;
    //     case 'X':
    //         cursor = data[rowIndex][colIndex];
    //         data[rowIndex][colIndex] = 'X';
    //         if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
    //         data[nextRowIndex][nextColIndex] = cursor;
    //         rowIndex = nextRowIndex;
    //         colIndex = nextColIndex;
    //         break;
    //     case '#':
    //         direction = (direction + 1) % 4;
    //         data[rowIndex][colIndex] = directions[direction];
    //         break;
    //     default:
    //         break;
    //     }
    //     if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
    // }
    // printf("Distinct tiles: %d\n", distinctTiles);
    // for (int i = 0; i < ctr; i++) {
    //     printf("%s\n", data[i]);
    // }
    int direction = 0;
    int distinctTiles = 0;
    char nextTile;
    char cursor;
    char cursorMask = 240;
    char visitedMask = 15;
    char directions[4] = { 16, 32, 64, 128 };
    char visitedDirection[4] = { 1, 2, 4, 8 };
    while (1) {
        int nextRowIndex = rowIndex;
        int nextColIndex = colIndex;
        char currentTile = data[rowIndex][colIndex] & cursorMask;
        switch (currentTile) {
        case 16:
            nextRowIndex--;
            break;
        case 32:
            nextColIndex++;
            break;
        case 64:
            nextRowIndex++;
            break;
        case 128:
            nextColIndex--;
            break;
        default:
            break;
        }

        if ((0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) {
            nextTile = data[nextRowIndex][nextColIndex] & visitedDirection;
        }
        switch (nextTile) {
        case '.':
            data[rowIndex][colIndex] = directionsMap[direction];
            if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
            data[nextRowIndex][nextColIndex] = currentTile | ;
            rowIndex = nextRowIndex;
            colIndex = nextColIndex;
            break;
        case '#':
            direction = (direction + 1) % 4;
            data[rowIndex][colIndex] = directions[direction];
            break;
        default:
            cursor = data[rowIndex][colIndex];
            data[rowIndex][colIndex] |= directionsMap[direction];
            if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
            data[nextRowIndex][nextColIndex] = cursor;
            rowIndex = nextRowIndex;
            colIndex = nextColIndex;
            break;
        }
        if (!(0 <= nextRowIndex && nextRowIndex < ctr && 0 <= nextColIndex && nextColIndex < strlen(data[rowIndex]))) break;
    }

    fclose(filePtr);
    return 0;
}